### ISSUE TITLE
Ignore order of fields in snapshot.

### ISSUE THREAD

**Issue by ftes** [json-snapshot/json-snapshot.github.io#9]
The snapshot is serialized as JSON. JSON does not guarantee ordering of object fields.

[RFC7159](https://tools.ietf.org/html/rfc7159)
>An object is an unordered collection of zero or more name/value pairs

However, the snapshot test fails if the order of fields in the serialized JSON has changed.

[Snapshot.java#L42](https://github.com/json-snapshot/json-snapshot.github.io/blob/master/src/main/java/io/github/jsonSnapshot/Snapshot.java#L42)
```java
if (!rawSnapshot.trim().equals(currentObject.trim()))
```

**Comment by andrebonna** [comment]
JSON is just a way to represent the object being tested with snapshot. If the objects compared are deeply equal, default GSON serializer will serialize them equally every time. If you provide a custom JSON serializer function, you must ensure the same behavior.

**Comment by tbo** [comment]
> default GSON serializer will serialize them equally every time

@andrebonna This doesn't match with my observations. The way some HashMaps are being serialized in my test cases seems to be non-deterministic. I get errors like this one all the time:

```
Changed content at line 22:
expecting:
  ["        "LFDNR": "1",",
   "        "KTOIBAN": "DE56940596610000008888""]
but was:
  ["        "KTOIBAN": "DE56940596610000008888",",
   "        "LFDNR": "1""]
```

Is there a way to avoid this behavior?

**Comment by andrebonna** [comment]
You are right, I totally forgot HashMaps serialization. Fortunatelly, Jackson serializer has a config property to solve this ordering issue, `SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS`. I just changed the default JSON serializer library from GSON to Jackson and used the config property to ensure always the same key ordering of a map. That should be fixed in version `1.0.14`. I'm reopening the issue until you confirm the bug has been solved on your side.

**Comment by tbo** [comment]
@andrebonna It seems to be fixed. Thanks!

**Comment by metas-ts** [comment]
Hi @andrebonna,

it still doesn't work for me, unless I also add
```java
objectMapper.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);
```

in the method `io.github.jsonSnapshot.SnapshotMatcher.buildObjectMapper()`

Further reading: https://www.stubbornjava.com/posts/creating-a-somewhat-deterministic-jackson-objectmapper

Pls let me know if I can help with this issue.

**Comment by andrebonna** [comment]
Provide us and example. JSON properties should be in the same order as Class properties definition.

**Comment by metas-ts** [comment]
> Please, provide us an example. JSON properties should be in the same order as Class properties definition.

Thx for the reply. I'm sorry, but I lack the time to look into trying to extract the case into an example. However, as we are open source, *if you want* you can 
* check out our code, 
* switch to branch ` 
* edit https://github.com/metasfresh/metasfresh-webui-api/blob/gh4959-app/pom.xml and remove the `-PR` from `<artifactId>json-snapshot-PR</artifactId>`
* and try out the test case https://github.com/metasfresh/metasfresh-webui-api/blob/gh4959-app/src/test/java/de/metas/ui/web/dataentry/window/descriptor/factory/DataEntryTabLoaderTest.java

* Without `objectMapper.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);` the `snap`file sas created on the first run, and then created differently on the second run
* When I run this with https://github.com/json-snapshot/json-snapshot.github.io/pull/12 plus `objectMapper.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);`, <br> then it worked **in Eclipse** (See next copmment for details, i believe it's related but not same same topic)

**Comment by metas-ts** [comment]
Turned out that for me even `objectMapper.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);` didn't really help to create the same snapshot on my local machine and our CI.
I suspect different java versions to be part of the problem 
 * Note: jackson was the same (tried both `2.8.8`and `2.9.8`), and we also use maven-enforcer to keep the dependency tree clean of duplicate classes and different versions of the same artifact)

The problem occurred with an interface that had default methods/properties:
* on my local machine the properties were still ordered alphabetically "globally"
* on jenkins I got the interface's properties (ordered alphabetically) followed by the implementing class's properties (also ordered alphabetically)

Here is my failed test: https://jenkins.metasfresh.com/job/metasfresh-webui/job/gh4959-app/12/testReport/de.metas.ui.web.dataentry.window.descriptor.factory/DataEntryTabLoaderTest/createGroupEntityDescriptors_verify_DocumentEntityDescriptor/ ..just for completeness..i'm not sure if it helps 😊 

I eventually got around this problem by explicitly setting the order (i.e. `@JsonPropertyOrder(value = { "columnName", "defaultOrderBy", "defaultOrderByAscending", "mandatory" })` ), but to me that's just a workaround. I want my collegues to love your tool and use it broadly, and I think for that, can'T  tell them "yeah, it's great but you need to instrument your production code like this and that etc etc".

So, I'm going to check out https://github.com/skyscreamer/JSONassert as documented here: https://www.baeldung.com/jsonassert .
I plan to extend `SnapshotConfig` to allow a more lenient comparison instead of the current `if (!rawSnapshot.trim().equals(currentObject.trim()))`.

**Comment by grimsa** [comment]
Maybe it would be possible to compare expected and actual JSON trees for equality using `JsonNode#equals`?
I'd expect Jackson to return true for two nodes that have the same fields and values, regardless of field order, but I have not checked.
